File: 0346-moving-average-from-data-stream/0346-moving-average-from-data-stream.rs
use std::collections::VecDeque;
struct MovingAverage {
    size : usize,
    movingAvg : VecDeque<i32>,
    sum : i64,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MovingAverage {

    fn new(size: i32) -> Self {
        MovingAverage {
            size : size as usize,
            movingAvg : VecDeque::with_capacity(size as usize),
            sum : 0,
        }
    }
    
    fn next(&mut self, val: i32) -> f64 {
        self.movingAvg.push_back(val);
        self.sum += val as i64;
        if self.movingAvg.len() > self.size{
            if let Some(removed) = self.movingAvg.pop_front(){
                self.sum -= removed as i64;
            }
        }
        self.sum as f64 / self.movingAvg.len() as f64
    }
}

/**
 * Your MovingAverage object will be instantiated and called as such:
 * let obj = MovingAverage::new(size);
 * let ret_1: f64 = obj.next(val);
 */

File: 0346-moving-average-from-data-stream/0346-moving-average-from-data-stream.java

/**
// a stream of integers, and a window size, 
calculate the moving avg of all integers in the sliding Window

- size : windowSize, return the moving avg of last size values of teh Stream

CQ: what would be the max size:?
In the sum / windowSize: , would sum > Integr.Max(32-bit)
How many calls will be made

Edge cases: 
1: window size 0, we return 0?
2: how do you want me to handle when the current avlue sin the stream is < windows Size

T: O(n): 
 */
class MovingAverage {
    int windowSize;
    Queue<Integer> stream;
    double runningSum = 0; // maybe we would not need this
    public MovingAverage(int size) {
        windowSize = size;
        stream = new LinkedList<>();
    }
    
    public double next(int val) {
        runningSum += val;
        double denominator = 1;
        stream.offer(val);
        if(stream.size() > windowSize){
            runningSum -= stream.poll(); //remove form the front of the queue
        }
        return runningSum / stream.size();
    }
}

/**
 * Your MovingAverage object will be instantiated and called as such:
 * MovingAverage obj = new MovingAverage(size);
 * double param_1 = obj.next(val);
 */